rts/Rendering/GL/glHelpers.h

```diff
@@ -0,0 +1,69 @@
+/* This file is part of the Recoil engine (GPL v2 or later), see LICENSE.html */
+
+#pragma once
+
+#include "myGL.h"
+#include <algorithm>
+#include <tuple>
+
+
+// Must pass expectedValuesN to convert from other integer types to GLint
+template<class GLType>
+inline void glGetAny(GLenum paramName, GLType* data, const int expectedValuesN = -1)
+{
+	GLint ints[expectedValuesN];
+	glGetIntegerv(paramName, ints);
+	std::move(ints, ints+expectedValuesN, data);
+}
+template<>
+inline void glGetAny<GLint>(GLenum paramName, GLint* data, const int)
+{
+	glGetIntegerv(paramName, data);
+}
+template<>
+inline void glGetAny<GLboolean>(GLenum paramName, GLboolean* data, const int)
+{
+	glGetBooleanv(paramName, data);
+}
+template<>
+inline void glGetAny<GLfloat>(GLenum paramName, GLfloat* data, const int)
+{
+	glGetFloatv(paramName, data);
+}
+template<>
+inline void glGetAny<GLdouble>(GLenum paramName, GLdouble* data, const int)
+{
+	glGetDoublev(paramName, data);
+}
+
+
+namespace GL
+{
+
+template<class ResultTupleType, GLenum... GLParamNames>
+inline ResultTupleType FetchActiveStateAttribValues()
+{
+	static constexpr GLenum GLParamNameFirst = std::get<0>(std::make_tuple(GLParamNames...));
+	static constexpr GLenum GLParamNameSecond = (sizeof...(GLParamNames) > 1? std::get<sizeof...(GLParamNames)-1>(std::make_tuple(GLParamNames...)) : 0);
+	static constexpr size_t ValuesPerParam = std::tuple_size_v<ResultTupleType>/sizeof...(GLParamNames);
+
+	ResultTupleType resultTuple;
+	glGetAny(GLParamNameFirst, &std::get<0>(resultTuple), ValuesPerParam);
+	if constexpr(GLParamNameSecond) glGetAny(GLParamNameSecond, &std::get<ValuesPerParam-1>(resultTuple), ValuesPerParam);
+	return resultTuple;
+}
+
+}
+
+template<auto DedicatedGLFuncPtrPtr, GLenum... GLParamName, class AttribValuesTupleType>
```

Comment:
Why mix `auto` and `class` in the declaration here? I think you can go with either (but preferably with `typename` as most of the rest of the codebase uses)

Reply:
_auto_ is an explicit template argument for glSetAny.
You use it like this:
glSetAny<&glDepthMask>(GL_TRUE) or glSetAny<nullptr, GL_DEPTH_TEST>(GL_TRUE).
_Any_ stands for whether it is a glEnable/Disable or named attribute, both can be set via this interface.
_class_ parameter is completely different, it is an automatically deduced template argument.